<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Describing set combinations with boxplot, tileplot and UpSet plot</title>
    <meta name="description" content="Phong BH. Nguyen's personal page - Scientist, programmer, adventurer. Learn about my research, projects, and interests.">
    <meta name="keywords" content="Phong BH. Nguyen, bioinformatics, data science, machine learning, health informatics, non-communicable diseases, diabetes, cancer, cardiovascular diseases">
    <meta name="author" content="Phong BH. Nguyen">
    <link type="application/atom+xml" rel="alternate" href="http://localhost:4001/feed.xml" />
    <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Describing set combinations with boxplot, tileplot and UpSet plot</title>
<meta name="generator" content="Jekyll v4.3.4" />
<meta property="og:title" content="Describing set combinations with boxplot, tileplot and UpSet plot" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="In this case I have a multiomic dataset, each omics has its own properties including number of features, number of samples and predictive performance for a phenotype. I was evaluating the performance of combinations of these omics and wanted to visualize the result. First I will give a descriptive vizualization of the dataset using UpSet plot and then prediction performance of the possible combinations using boxplot and tileplot." />
<meta property="og:description" content="In this case I have a multiomic dataset, each omics has its own properties including number of features, number of samples and predictive performance for a phenotype. I was evaluating the performance of combinations of these omics and wanted to visualize the result. First I will give a descriptive vizualization of the dataset using UpSet plot and then prediction performance of the possible combinations using boxplot and tileplot." />
<link rel="canonical" href="http://localhost:4001/data_viz/multidims_boxplot/" />
<meta property="og:url" content="http://localhost:4001/data_viz/multidims_boxplot/" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-01-30T12:10:33+01:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Describing set combinations with boxplot, tileplot and UpSet plot" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","dateModified":"2025-01-30T12:10:33+01:00","datePublished":"2025-01-30T12:10:33+01:00","description":"In this case I have a multiomic dataset, each omics has its own properties including number of features, number of samples and predictive performance for a phenotype. I was evaluating the performance of combinations of these omics and wanted to visualize the result. First I will give a descriptive vizualization of the dataset using UpSet plot and then prediction performance of the possible combinations using boxplot and tileplot.","headline":"Describing set combinations with boxplot, tileplot and UpSet plot","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4001/data_viz/multidims_boxplot/"},"url":"http://localhost:4001/data_viz/multidims_boxplot/"}</script>
<!-- End Jekyll SEO tag -->

    <link href="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/themes/prism-tomorrow.min.css" rel="stylesheet" />
    <style>
      body {
        background: rgba(64,64,64,0.8);
        font-family: Arial, sans-serif;
        margin: 0;
        padding: 0;
        display: flex;
        flex-direction: column;
        height: 100vh;
        color: white;
        font-size: 2vw;
      }
      .tabs {
        display: flex;
        justify-content: center;
        align-items: center;
        width: 100%;
        padding: 0.8vw;
        background: rgba(43, 43, 43, 1); /* Coffee color with reduced opacity */
        font-size: 3vw;
        position: fixed;
      }
      .tabs a {
        color: #F7CE46;
        text-decoration: none;
        margin: 0 10px;
        padding: 10px;
        background: none;
        border-radius: 1vw;
        font-size: 1.3vw;
      }
      .tabs a.current {
        background: #F7CE46; /* Different color for the current page */
        color: rgba(43, 43, 43, 1);
        border-radius: 1vw;
        font-size: 1.5vw;
      }
      .menu-button {
        display: none; /* Hide by default */
      }
      .toc-button {
        display: none; /* Hide by default */
      }
      .container {
        display: flex;
        flex: 1;
      }
      .left-bar {
        background: rgba(51, 51, 51, 0.9); /* Coffee color with reduced opacity */
        width: 16vw;
        padding: 0.5vw;
        position: fixed;
        top: 5vw;
        display: flex;
        flex-direction: column;
        align-items: center;
        font-size: 2.5vw;
      }
      .left-bar img {
        border-radius: 50%;
        width: 10vw;
        height: 10vw;
        object-fit: cover;
        margin-bottom: 0.5vw;
        margin-left: 0.8vw;
      }
      .left-bar ul {
        list-style: none;
        padding: 0;
      }
      .left-bar ul li {
        margin: 0.3vw 0.3vw;
      }
      .left-bar ul li a {
        color: rgb(221,221,221);
        text-decoration: none;
        font-size: 1.2vw;
      }
      .content {
        flex: 1;
        padding: 2vw;
        background: rgba(64,64,64,0.8); /* Different color for the content area */
        color: rgb(221,221,221);
        font-size: 1vw;
        margin-left: 230px;
        margin-top: 40px;
        text-align: justify;
      }
      .content a {
        color: #F7CE46; /* Change hyperlink color */
      }
      .bash-command {
          font-family: monospace;
          background: #272727;
          color: rgb(221,221,221);
          padding: 0px;
          margin: 0px;
          border-radius: 5px;
          overflow-x: auto;
          white-space: pre;
          font-size: 1vw;
          display: block;
          width: 1100px;
      }
      pre[class*="language-"] {
        background: #272727;
        padding: 0px;
        margin: 0px;
        border-radius: 5px;
        overflow-x: auto;
        display: block;
        width: 1100px;
      }
      /* Custom styles for headers */
      h1, h2, h3, h4, h5, h6 {
        color: #F7CE46; /* Change header color */
        background-color: rgba(43, 43, 43, 0.8); /* Highlight background color */
        padding: 5px; /* Add padding */
        border-radius: 5px; /* Add border radius */
      }
      .thumbnail {
        width: 15%;
        float: left;
        margin-right: 15px;
        margin-bottom: 15px;
        height: auto;
      }
      .project-item {
        clear: both;
        margin-bottom: 30px;
      }
      @media (max-aspect-ratio: 1/1) {
        body {
          font-size: 2vw; /* Adjust font size for smaller screens */
        }
        .tabs {
          justify-content: space-between;
          padding: 0px;
          font-size: 8px; /* Adjust font size for smaller screens */
        }
        .tabs a {
          display: none; /* Hide tabs on phone screens */
        }
        .tabs a.current {
          font-size: 4vw;
        }
        .menu-button {
          display: block; /* Show menu button on phone screens */
          background: #F7CE46;
          color: rgba(43, 43, 43, 1);
          border: none;
          padding: 5px;
          font-size: 4vw;
          cursor: pointer;
          position: fixed;
          border-radius: 5px;
          top: 8px;
          right: 10px;
        }
        .menu-content {
          display: none;
          position: absolute;
          top: 40px;
          right: 20px;
          background: rgba(43, 43, 43, 1);
          padding: 5px;
          border-radius: 5px;
          z-index: 1001;
        }
        .menu-content a {
          display: block;
          color: #F7CE46;
          text-decoration: none;
          margin: 5px 0;
          padding: 5px;
          font-size: 4vw;
        }
        .left-bar {
          width: 16%; /* Adjust width for smaller screens */
          padding: 4px;
          font-size: 2vw; /* Adjust font size for smaller screens */
          top: 7px;
          border-radius: 3px;
        }
        .left-bar img {
          width: 8vw; /* Smaller image for smaller screens */
          height: 8vw;
          margin-bottom: 2px;
          margin-left: 5px;
        }
        .left-bar ul li {
          margin: 5px 0px;
        }
        .left-bar ul li a {
          font-size: 2vw; /* Adjust font size for smaller screens */
          padding: 2px;
          margin-top: 2px;
        }
        .toc-button {
          display: block; /* Show table of content button on phone screens */
          background: #F7CE46;
          color: rgba(43, 43, 43, 1);
          border: none;
          padding: 3px 3px;
          font-size: 2vw;
          cursor: pointer;
          border-radius: 3px;
          margin-top: 3px;
        }
        .toc-content {
          display: none;
          background: rgba(51, 51, 51, 0.9);
          color: #F7CE46;
          padding: 8px;
          border-radius: 5px;
          margin-left: 5px;
          width: 100%;
          font-size: 3vw;
        }
        .toc-content a {
          display: block;
          text-decoration: none;
          margin: 5px 5px;
          padding: 5px;
          font-size: 3vw;
        }
        .content {
          padding: 3vw;
          font-size: 3vw; /* Adjust font size for smaller screens */
          margin-top: 0px;
          margin-left: 15vw;
        }
        .bash-command {
          font-family: monospace;
          background: #272727;
          color: rgb(221,221,221);
          padding: 0px;
          border-radius: 5px;
          overflow-x: auto;
          white-space: pre;
          font-size: 8px;
          display: block;
          width: 330px;
          margin: 0px;
        }
        .thumbnail {
          width: 50%;
        }
      }
    </style>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/prism.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/components/prism-bash.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/components/prism-perl.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/components/prism-r.min.js"></script>
    <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
  </head>
  <body>
    <div class="tabs">
      <button class="menu-button" onclick="toggleMenu()">Menu</button>
      <div class="menu-content" id="menu-content">
        <nav>
  
    <a href="/">Home</a>
  
    <a href="/about.html">About me</a>
  
    <a href="/projects.html">Projects/ tools development</a>
  
    <a href="/data_viz.html">Data visualization</a>
  
    <a href="/blog.html">Blog</a>
  
</nav>
      </div>
    </div>
    <div class="container">
      <div class="left-bar">
        <img src="/assets/images/page_img.jpg" alt="Profile Picture">
        <button class="toc-button" onclick="toggleToc()">Table of Content</button>
        <div class="toc-content" id="toc-content">
          <ul>
            
              <li><a href="#chapter1">Introduction</a></li>
            
              <li><a href="#chapter2">UpSet plot</a></li>
            
              <li><a href="#chapter3">Boxplot and Tileplot</a></li>
            
          </ul>
        </div>
      </div>
      <div class="content">
        <p><a id="chapter1"></a>
In this case I have a multiomic dataset, each omics has its own properties including number of features, number of samples and predictive performance for a phenotype. I was evaluating the performance of combinations of these omics and wanted to visualize the result. First I will give a descriptive vizualization of the dataset using UpSet plot and then prediction performance of the possible combinations using boxplot and tileplot.</p>

<p>First we need to load necessary R libraries.</p>
<div class="bash-command">
    <pre><code class="language-R">
        library(dplyr)
        library(ggplot2)
        library(ComplexHeatmap)
        library(RColorBrewer)
    </code></pre>
</div>

<h2 id="upset-plot">UpSet plot</h2>
<p><a id="chapter2"></a></p>

<p>UpSet plot is useful to visualize set data with many intersecting sets, such as our multiomic dataset. The list of all samples in the multiomic dataset is provided as a dataframe of sample IDs. Each column represent one omic. IDs that belong to the same sample are in the same row, missing data is shown as <code class="language-plaintext highlighter-rouge">NA</code>.</p>

<div style="text-align: center;">
  <img src="/assets/images/multidims_boxplot_fig1.png" alt="Figure 1" style="max-width: 70%; height: auto;" />
  <p style="max-width: 80%; margin: auto; font-size: 70%;"><strong>Dataframe of sample IDs for the multiomic dataset.</strong></p> 
</div>

<p>Then go a head and make the UpSet plot:</p>

<div class="bash-command">
    <pre><code class="language-R">
        # Extract samples of each omic and represent with a single ID type (Clinical)
        gen &lt;- as.character(omics_id$Clinical)[!is.na(omics_id$Genomics)]
        tra &lt;- as.character(omics_id$Clinical)[!is.na(omics_id$Transcriptomics)]
        pro &lt;- as.character(omics_id$Clinical)[!is.na(omics_id$Proteomics)]
        meta &lt;- as.character(omics_id$Clinical)[!is.na(omics_id$Metabolomics)]
        meth &lt;- as.character(omics_id$Clinical)[!is.na(omics_id$Methylomics)]
        oli &lt;- as.character(omics_id$Clinical)[!is.na(omics_id$Olink)]

        # Put them into a list
        l &lt;- list(Genomics = gen, 
                Transcriptomics = tra, 
                Proteomics = pro, 
                Metabolomics = meta, 
                Methylomics = meth, 
                ClinicalData_InflamatoryProteins = oli)

        # Make a combination matrix 
        m &lt;- make_comb_mat(l, mode = "intersect") # Make intersections of sets

        # Set a color scheme
        color_types &lt;- brewer.pal(7,"Set2")
        names(color_types) &lt;- c("Genomics",
                                "Transcriptomics",
                                "Proteomics",
                                "Metabolomics",
                                "Methylomics",
                                "Olink",
                                "Clinical")

        # Make UpSet plot
        UpSet(m,
            set_order = names(l), 
            comb_order = order(comb_size(m)), # Order based on size of the sets
            right_annotation = upset_right_annotation(m, # Setting for annotation of the sets
                            add_numbers = TRUE, 
                            gp = gpar(fill = color_types, col = "white")),
            column_names_rot = 45, 
            pt_size = unit(1,"mm"), 
            lwd = 1)

    </code></pre>
</div>

<p>Here is the result:</p>

<div style="text-align: center;">
  <img src="/assets/images/multidims_boxplot_fig2.png" alt="Figure 1" style="max-width: 90%; height: auto;" />
  <p style="max-width: 80%; margin: auto; font-size: 70%;"><strong>Upset plot of the sample distribution among datasets.</strong></p> 
</div>

<p>Documentation of how to adjust other parameters of the UpSet plot and other functions of the package <code class="language-plaintext highlighter-rouge">ComplexHeatmap</code> could read <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">here</a>.</p>

<h2 id="boxplot-and-tileplot">Boxplot and Tileplot</h2>
<p><a id="chapter3"></a></p>

<p>Next I would like to visualize the prediction performance of different combinations of omics as features of the model. After the model training, I got this dataframe as result:</p>

<div style="text-align: center;">
  <img src="/assets/images/multidims_boxplot_fig3.png" alt="Figure 3" style="max-width: 70%; height: auto;" />
  <p style="max-width: 80%; margin: auto; font-size: 70%;"><strong>Prediction performance of the multiomic models.</strong></p> 
</div>

<p>where each column shows result of a cross-validation iteration:</p>
<ul>
  <li>cv: 10-fold cross-validation, one iteration per row</li>
  <li>phenotype: the target variable</li>
  <li>omics: single omics or combinations of omics</li>
  <li>sens, spec, auc: performance metrics, sensitivity, specificity and AUROC respectively</li>
  <li>type: either ‘classification’ or ‘regression’</li>
  <li>uselab: whether I used laboratory measurement as features or not</li>
</ul>

<p>Here the dataframe was named <code class="language-plaintext highlighter-rouge">cv</code> in the R environment. The goal is to show the evaluated combinations of omics and the prediction performance for each combination at the same time. To this end I used tileplot to make the former and boxplot to make the latter, then used the <code class="language-plaintext highlighter-rouge">cowplot</code> package to combine the two with matched coordinate. Here is details how to do it:</p>

<div class="bash-command">
    <pre><code class="language-R">
        # Rank the combinations based on median AUROC
        cv_tmp &lt;- filter(cv, uselab != "no") %&gt;% # Only consider iterations that 'uselab'
            dplyr::select(omics, auc) %&gt;% 
            group_by(omics) %&gt;%
            summarise(med = median(auc, na.rm = T)) %&gt;% 
            arrange(.,desc(med)) %&gt;% # Descending order 
            mutate(rank = 1:nrow(.)) # Add ranking

        # Append a zeros matrix with each column represents one omics
        cv_tmp &lt;- cv_tmp %&gt;%
        mutate(Genomics = vector(mode = "numeric", length = nrow(cv_tmp)),
                Transcriptomics = vector(mode = "numeric", length = nrow(cv_tmp)),
                Proteomics = vector(mode = "numeric", length = nrow(cv_tmp)),
                Metabolomics = vector(mode = "numeric", length = nrow(cv_tmp)),
                Methylomics = vector(mode = "numeric", length = nrow(cv_tmp)),
                Olink = vector(mode = "numeric", length = nrow(cv_tmp)),
                Covariate = vector(mode = "numeric", length = nrow(cv_tmp)))

        # Transform the zeros matrix into a binary matrix showing whether the omics presents in the combination or net
        for (i in 1:nrow(cv_tmp)){
            cv_tmp$Genomics[i] &lt;- ifelse(length(grep("Genomics",cv_tmp$omics[i])) &gt; 0, 1, 0)
            cv_tmp$Transcriptomics[i] &lt;- ifelse(length(grep("Transcriptomics",cv_tmp$omics[i])) &gt; 0, 1, 0)
            cv_tmp$Proteomics[i] &lt;- ifelse(length(grep("Proteomics",cv_tmp$omics[i])) &gt; 0, 1, 0)
            cv_tmp$Metabolomics[i] &lt;- ifelse(length(grep("Metabolomics",cv_tmp$omics[i])) &gt; 0, 1, 0)
            cv_tmp$Methylomics[i] &lt;- ifelse(length(grep("Methylomics",cv_tmp$omics[i])) &gt; 0, 1, 0)
            cv_tmp$Olink[i] &lt;- ifelse(length(grep("Olink",cv_tmp$omics[i])) &gt; 0, 1, 0)
            cv_tmp$Covariate[i] &lt;- 1
        }

        # Transform into a long table
        cv_tmp &lt;- pivot_longer(cv_tmp, cols = Genomics:Covariate, names_to = "Type") 
        cv_tmp$size &lt;- rep(300, each = nrow(cv_tmp))

        # Add the ranking to the CV dataframe
        cv$rank &lt;- cv_tmp$rank[match(cv$omics,cv_tmp$omics)]

        # Plot AUROC
        p1 &lt;- ggplot(cv, 
                    aes(x = reorder(omics, desc(rank)), 
                    y = auc, 
                    fill = uselab, 
                    col = uselab))+
                geom_boxplot(position = "identity", alpha = 0.5) +
                geom_hline(yintercept = 0.5, col = "blue") +
                xlab("Omics") +
                ylab("AUROC") +
                theme(legend.position = "left", axis.text.y = element_blank()) +
                coord_flip() 
        # Plot sensitivity
        p2 &lt;- ggplot(cv, 
                    aes(x = reorder(omics, desc(rank)), 
                    y = sens, 
                    fill = uselab, 
                    col = uselab))+
                geom_boxplot(position = "identity", alpha = 0.5) +
                xlab("Omics") +
                ylab("Sensitivity") +
                theme(legend.position = "none", axis.title.y = element_blank(),axis.text.y = element_blank()) +
                coord_flip() 
        # Plot specificity
        p3 &lt;- ggplot(cv, 
                    aes(x = reorder(omics, desc(rank)), 
                    y = spec, 
                    fill = uselab, 
                    col = uselab))+
            geom_boxplot(position = "identity", alpha = 0.5) +
            ylab("Specificity") +
            theme(legend.position = "none", axis.title.y = element_blank(), axis.text.y = element_blank()) +
            coord_flip() 
        # List of combinations
        o_list &lt;- unique(cv_tmp$omics) %&gt;% as.character()
        # Plot the combination names
        p4 &lt;- ggplot(cv_tmp) + 
                geom_tile(aes(x=Type,
                        y=reorder(omics,desc(rank)), 
                        fill = factor(value)), 
                        colour="white", 
                        size=0.5) + 
                geom_text(aes(x=Type,
                        y=reorder(omics,desc(rank)),
                        label=ifelse(value==1, substr(Type,1,4), "")), colour="white", 
                        size=2) +
                theme(axis.title =element_blank(),
                    axis.text.y = element_blank(),
                    axis.text.x=element_text(angle=-90, hjust=0, vjust=0.5)
                ) +
                scale_fill_manual(values=c("grey90", "grey30")) +
                geom_rect(data = filter(cv_tmp, omics == "MetabolomicsOlink"),
                            aes(xmin=-Inf, xmax=Inf, ymin=nlevels(cv_tmp$omics) - which(o_list == "ProteomicsMetabolomicsOlink")-0.5 + 1,
                            ymax=nlevels(cv_tmp$omics) - which(o_list=="ProteomicsMetabolomicsOlink")+0.5 + 1),
                            fill = "red", alpha = 0.05)+ # Label the best performing combination
                guides(fill=FALSE, alpha = FALSE, col = F)
        
        # Put everything together
        ggarrange(p1,p2,p3,p4, ncol = 4, align = "hv", widths = c(2,2,2,1.5))

    </code></pre>
</div>

<p>And here is the result:</p>

<div style="text-align: center;">
  <img src="/assets/images/multidims_boxplot_fig4.png" alt="Figure 4" style="max-width: 70%; height: auto;" />
  <p style="max-width: 80%; margin: auto; font-size: 70%;"><strong>Prediction performance of the multiomic models.</strong></p> 
</div>

<p>The plot nicely show the combinations of omics that I evaluated using the tileplot on the right and their corresponding prediction performance using the boxplots on the left.</p>


      </div>
    </div>
    <script>
      function toggleMenu() {
        var menuContent = document.getElementById('menu-content');
        if (menuContent.style.display === 'block') {
          menuContent.style.display = 'none';
        } else {
          menuContent.style.display = 'block';
        }
      }
      // Hide menu when clicking outside of it
      document.addEventListener('click', function(event) {
        var menu = document.getElementById('menu-content');
        var menuButton = document.querySelector('.menu-button');
        if (menu.style.display === 'block' && !menu.contains(event.target) && !menuButton.contains(event.target)) {
            menu.style.display = 'none';
        }
      });
      function toggleToc() {
        var tocContent = document.getElementById('toc-content');
        if (tocContent.style.display === 'block') {
          tocContent.style.display = 'none';
        } else {
          tocContent.style.display = 'block';
        }
      }
      // Hide menu when clicking outside of it
      document.addEventListener('click', function(event) {
        var menu = document.getElementById('toc-content');
        var menuButton = document.querySelector('.toc-button');
        if (menu.style.display === 'block' && !menu.contains(event.target) && !menuButton.contains(event.target)) {
            menu.style.display = 'none';
        }
      });
    </script>
  </body>
</html>